### PR TITLE
KAFKA-3937: Kafka Clients Leak Native Memory For Longer Than Needed With Compressed Messages

### PR DESCRIPTION
**Pull request by wiyu**
@ijuma - Creating this PR against 0.9.0 as this is what we're using in prod. I can modify this for trunk if the logic solution looks good. 

I modified the record iterator to inherit from closable and implemented the method to call streams.close(). The stream will now be closed when we are done traversing the inner iterator. I had to move the `try {} catch{}` out of the `if` portion to deal with IOException coming from `innerDone()`.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java [hunk 1]
```diff
@@ -271,18 +272,43 @@ protected LogEntry makeNext() {
...
                 }
-            } else {
-                return innerIter.next();
+                return allDone();
+            } catch (IOException e) {
+                try {
+                    close();
+                } catch (IOException ioe) {
+                    throw new KafkaException(ioe);
+                }
```
Comment (wiyu) [inline clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java:291]: Losing the original exception here if the stream fails to close. Any thoughts on the best way to handle these potential stream closed exceptions from within a `catch` block?

**Comment by ijuma** [comment]
I suggest doing this on `trunk` where we can follow the same approach we did for the server. That is, we remove the special-case for V0 below and close the stream in a finally.

``` java
if (entry.record().magic() > Record.MAGIC_VALUE_V0) {
                this.logEntries = new ArrayDeque<>();
                long wrapperRecordTimestamp = entry.record().timestamp();
                while (true) {
                    try {
                        LogEntry logEntry = getNextEntryFromStream();
                        Record recordWithTimestamp = new Record(logEntry.record().buffer(),
                                                                wrapperRecordTimestamp,
                                                                entry.record().timestampType());
                        logEntries.add(new LogEntry(logEntry.offset(), recordWithTimestamp));
                    } catch (EOFException e) {
                        break;
                    } catch (IOException e) {
                        throw new KafkaException(e);
                    }
                }
                this.absoluteBaseOffset = wrapperRecordOffset - logEntries.getLast().offset();
```

**Comment by wiyu** [comment]
@ijuma sure, I can do that, but how would the change go back into 0.9.0 or is it no longer updated?  Do I need to self patch and release my application with the forked changes?

Is this what you had in mind about the change about ignoring the V0 and using finally?

```
            try {
                if ( entry.record().magic() > Record.MAGIC_VALUE_V0 ) {
                    this.logEntries = new ArrayDeque<>();
                    long wrapperRecordTimestamp = entry.record().timestamp();
                    while ( true ) {
                        try {
                            LogEntry logEntry = getNextEntryFromStream();
                            Record recordWithTimestamp = new Record(
                                logEntry.record().buffer(),
                                wrapperRecordTimestamp,
                                entry.record().timestampType()
                            );
                            logEntries.add(new LogEntry(logEntry.offset(), recordWithTimestamp));
                        }
                        catch (EOFException e) {
                            break;
                        }
                        catch (IOException e) {
                            throw new KafkaException(e);
                        }
                    }
                    this.absoluteBaseOffset = wrapperRecordOffset - logEntries.getLast().offset();
                }
                else {
                    this.logEntries = null;
                    this.absoluteBaseOffset = -1;
                }
            } finally {
                try {
                    stream.close();
                } catch (IOException e) {
                    throw new KafkaException(e);
                }
            }
```

**Comment by ijuma** [comment]
@wiyu, if we want to make the changes to `0.9.0`, then they will have to be different as you suggested. The reason why I prefer the approach I suggested is that it handles the case where the iterator is not consumed fully as well. But I haven't looked in detail, so if it's possible for the same solution to handle that in 0.9.0, 0.10.0 and trunk, then we can consider that.

With regards to what I meant, here's the equivalent for the server code:

https://github.com/apache/kafka/pull/1660/files

**Comment by wiyu** [comment]
Closing, will open a new PR.

**Pull request closed without merging**